tasks/base_tasks.py

```diff
@@ -1592,3 +1592,127 @@ def collect_meta_wrappers(test_module=None, test_all=True):
+        return PostgresTarget(schema=self.schema, tablename=self.table, non_empty=False)
+
+
+def base_downloader(url, output_path):
+    urllib.request.urlretrieve(url, output_path)
+
+
+class RepoFile(Task):
+    resource_id = Parameter()
+    url = Parameter()
+    version = IntParameter()
+    downloader = Parameter(default=base_downloader)
+
+    _repo_dir = 'repo'
+    _path = None
+
+    def requires(self):
+        return CreateRepoTable()
+
+    def run(self):
+        self.output().makedirs()
+        self._retrieve_remote_file()
+        digest = self._digest(self.output().path)
+        self._to_db(self.resource_id, self.version, digest, self.url, self.output().path)
+
+    def _create_filepath(self):
+        return self._build_path(str(uuid.uuid4()))
+
+    def _build_path(self, filename):
+        return os.path.join(self._repo_dir, self.resource_id, str(self.version), filename)
+
+    def _retrieve_remote_file(self):
+        self.downloader(self.url, self.output().path)
+
+    # https://stackoverflow.com/questions/3431825/generating-an-md5-checksum-of-a-file
+    def _digest(self, file):
+        hash_md5 = hashlib.md5()
+        with open(file, "rb") as f:
+            for chunk in iter(lambda: f.read(4096), b''):
+                hash_md5.update(chunk)
+        return hash_md5.hexdigest()
+
+    def _from_db(self, resource_id, version):
+        checksum, url, path = None, None, None
+        query = '''
+                SELECT checksum, url, path FROM "{schema}".{table}
+                WHERE id = '{resource_id}'
+                  AND version = {version}
+                '''.format(schema=self.input().schema,
+                           table=self.input().tablename,
+                           resource_id=resource_id,
+                           version=version)
+        result = current_session().execute(query).fetchone()
+        if result:
+            checksum, url, path = result
+
+        return checksum, url, path
+
+    def _to_db(self, resource_id, version, checksum, url, path):
+        query = '''
+                INSERT INTO "{schema}".{table}
+                    (id, version, url, checksum, path, added)
+                SELECT '{resource_id}', {version}, '{url}', '{checksum}', '{path}', now();
+                '''.format(schema=self.input().schema,
+                           table=self.input().tablename,
+                           resource_id=resource_id,
+                           version=version,
+                           url=url,
+                           checksum=checksum,
+                           path=path)
+        current_session().execute(query)
+        current_session().commit()
+
+    def _get_filepath(self):
+        path = self._from_db(self.resource_id, self.version)[2]
+        if not path:
+            if not self._path:
+                self._path = self._create_filepath()
+            path = self._path
+
+        return path
+
+    def complete(self):
+        deps = self.deps()
```

Comment:
Hate this too but unless we find a way to get rows from a table we are stick to this kind of `complete` methods which are error-prone